articles/qos_configurability.md

```diff
@@ -0,0 +1,374 @@
+
+Not providing an standard mechanism for reconfiguring QoS results in a lot of slightly different interfaces to do the same thing, creating a pain point for end users when figuring out how to reconfigure QoS for nodes in their system.
+
+## Design goals
+
+This proposal was written with the following requirements in mind:
+
+- Shall be opt-in on a per entity basis.
+  - Rationale: Allowing to change all QoS policies could break a contract the node's implementer assumed.
+  For example, the node's implementer may have assumed that the data was durable and not volatile.
+  Allowing the node's author to only accept some reconfigurations will avoid user errors.
+- Shall allow the developer to accept/reject changes to a QoS profile with a validation callback.
+  - See rationale above.
+- Shall only apply to publisher/subscriber QoS.
+  - Rationale: Though you can currently pass a custom QoS profiles to services or actions, configuring them is quite fragile and some of the policies cannot be modified (e.g. best effort services don't work).
+  Until a better QoS profile for services and actions is designed, this feature won't be exposed.
+- Shall only allow changing QoS at startup time, and not during runtime by default.
+  - Rationale: Allowing runtime reconfiguration of QoS is a complex topic, and purposefully left out of scope in this proposal.
+- Shall have a way of reusing the same QoS profile.
+  - Rationale: Not having a way of setting the same QoS profile to different entities will make configuring entities which profiles must match (or be compatible) error prone.
+
+## Adding a new mechanism vs using parameters
+
+A previous proposal (see PR [#296](https://github.com/ros2/design/pull/296)) suggested using a new file format for reconfiguring QoS policies.
+Some concerns were raised by the community:
+1. It adds a different mechanism to the ones that already exist.
+    Newcomers have to learn yet another thing.
+1. QoS settings are not conceptually different to other parameters, e.g.: publishing rate.
+1. It adds a new mechanism to configure QoS profiles.
+
+Based on that feedback, this PR will explore a design based on parameters.
+
+## Introduction
+
+Current code will not allow QoS reconfigurability if not changed:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10});
+```
+
+To make reconfigurability easy, only a flag is needed to automatically create the parameters:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10},
+  QosOverridingOptions{true});  // allow_reconfigurable all qos
+```
+
+That will automatically declare the parameters for reconfiguring the QoS policies that can be overridden in the following way:
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:  # group for all the qos settings
+      publisher:
+        'my/fully/qualified/topic/name/here': 
```

Comment:
In the most common use cases, each topic has only one message type.

Always including the topic type in the parameter sounds a bit verbose to me, and the "publisher/subscription id" trick mentioned later should cover this use case.
Anyways, including the topic type in the parameter name sounds like an acceptable proposal to me (though it will get problematic to make the parameters URI unique for a given fq node name, entity type, fq topic name and topic type).

Reply:
While it's true that each topic _typically_ has only one associated message type, I don't think we should build in this assumption since, in general, ROS2 allows for multiple types per topic. Regardless of particular solution, there should be a way to differentiate between TopicA/MessageTypeA and TopicA/MessageTypeB when specifying QoS settings.

Reply:
> I'm wondering about using the ID approach (when creating a publisher)... I suppose this change would propagate to other functions, such as get_publishers_info_by_topic

I don't think that propagating the concept to graph functions is a good idea.

Having an ad-hoc id only to be able to distinguish publishers/subscribers when re-configuring QoS might not be a great idea either.
If we think that distinguishing publishers in the same topic within a node is not necessary (see [this comment](https://github.com/ros2/design/pull/300#discussion_r499884787)), the concept of "id" can be deleted and the topic type can be added to the parameter name.

It's a bit problematic to add the topic type to the parameter URI though, e.g.:

```yaml
  ros__parameters:
    qos_profiles:  # group for all the qos settings
      publisher:
        'my/fully/qualified/topic/name/here':
          'std_msgs.msg.String':
```

current addressing scheme:
> ros_param://my/full/namespace/ny_node_name/qos_profiles.publisher.my/fully/qualified/topic/name/here.std_msgs.msg.String

alternative addressing scheme
> ros_param://my.full.namespace.ny_node_name/qos_profiles/publisher/my/fully/qualified/topic/name/here.std_msgs.msg.String

well, that actually seems to work.

---

That will make the parameter names much more verbose, but I think that's ok.
The type cannot be optional, because that would imply declaring two parameters (except if we want to not use parameters for this, like in https://github.com/ros2/design/pull/296).

Reply:
Yes, it is more verbose, but I think it's consistent with the general idea of allowing for multiple types per topic name. 

Regarding this:
> If we think that distinguishing publishers in the same topic within a node is not necessary...

I would recommend starting with an addressing scheme that includes the type. Then, see if someone comes up with the requirement to distinguish between multiple publishers with the same topic and type in a single node. I'd like to hear what others have to say about this, though.

Reply:
By the way, I see from reading up a bit on DDS that a "topic" is defined as a triad composed of a name, a type, and a set of QoS policies (I've been a little sloppy in limiting the "topic" term to the name; I'll try to be more precise). If we're trying to assign QoS policies for topics, it would make sense to me that we allow the user to specify the name and type to which they apply.

Reply:
> Yes, it is more verbose, but I think it's consistent with the general idea of allowing for multiple types per topic name.

I'm not sure if we guarantee that multiple types in a topic will work, or it's only working because the underlying middleware is DDS.
I think we don't miss any use case by requesting to have only one type per topic name, you can easily remap the colliding names to something else.

> I'd like to hear what others have to say about this, though.

:+1:

---

> on DDS that a "topic" is defined as a triad composed of a name, a type, and a set of QoS policies

A bit off topic, but I wouldn't consider the topic definition to be a triad.
e.g.: DW=data writer, DR=data reader

- DW A: reliable + defaults
- DW B: best effort + defaults
- DR A: reliable + defaults
- DR B: best effort + defaults
- All DW and DR use the same topic name and type.

DW A will match DR A and DR B.
DW B will match DR B.

If QoS are considered part of the topic definition, the above example is a bit strange.

IMHO, the easiest way to see it is that a name and a type defines a topic, but that doesn't guarantee that the endpoints will match. The DW offered QoS must satisfy the DR requested QoS in order to communicate.

P.S.: DDS has "Topic QoS", but those are used as a default for DR/DW and don't affect communication (but, can also be shared during discovery).

Reply:
@ivanpauno Good, thanks for the background info.

Reply:
@gbiggs  One example is the ROS2 node's get_topic_names_and_types method, which returns a map of strings to vectors of strings:

```
  /// Return a map of existing topic names to list of topic types.
  /**
   * \sa rclcpp::Node::get_topic_names_and_types
   */
  RCLCPP_LIFECYCLE_PUBLIC
  std::map<std::string, std::vector<std::string>>
  get_topic_names_and_types(bool no_demangle = false) const;
```
Implying that each topic name can be associated with multiple types. 

I don't want to get off-topic, but if this is indeed the case I think this also will break any tools that make the assumption of one topic name per one topic type.  For example, rqt_bag, which I've been working on, has a row on the message timeline for each topic name. To properly support 1 topic name/n topic types, rqt_bag would need one row per combination.

Reply:
My personal opinion after all of this is that if you use more than one type per topic, and you want to configure these, then you should use a named entity (so called "id").

---

I think it's somewhat unlikely that within a single node there are two entities with the same topic name but separate types. I would consider this a bug and discourage it. It is more likely, in my opinion, that two entities in different nodes have the same name but different types, even though I'd still discourage that. But in that situation you could still configure them separately as they are namespaced by the node name and namespace.

It is true that we allow for more than one type per topic, but that's not because we encourage it, but because of the nature of distributed discovery, since there might be remote entities with the same name but different types. That's the main reason for the discovery API allowing for that.

In fact, unless it changed, most DDS vendors will not allow you to create a topic with the same name but different types within a single participant. It only comes up when the conflicting entity is remote or at least in a different participant. So their behavior is a bit asymmetric on that point.

I think adding the type to the name of the parameter could work, but I'd prefer keeping the common case simple (each topic has a different type) and letting users use the named entity mechanism if they need to do the more complicated thing (entities share a topic name but not a type).

Reply:
> My personal opinion after all of this is that if you use more than one type per topic, and you want to configure these, then you should use a named entity (so called "id").

My last question about this: Do we want the name of the entity ("id") to be complementary to the topic name (like described in the [current proposal]) or to be a replacement for it (i.e. the default "id" is the topic name and that can be overridden)?

More details in [this comment](https://github.com/ros2/design/pull/300#issuecomment-705123339).

I don't have a strong preference about either option, using the "id" as an optional complement to the topic name is more restrictive, but that can be seen as an advantage (less ways of people getting creative on how to provide qos profile overrides).

[current proposal]: https://github.com/ros2/design/blob/d4558599c9a744e79715d4a1d0b4d185b41005bf/articles/qos_configurability.md

Reply:
I don't have any opinion on `id` vs `<entity>_id`.